packages/nightingale-sunburst/src/nightingale-sunburst.js

```diff
@@ -0,0 +1,358 @@
+import { LitElement, html } from "lit-element";
+import {
+  select,
+  partition as d3partition,
+  hierarchy,
+  scaleOrdinal,
+  quantize,
+  interpolateRainbow,
+} from "d3";
+
+const superkingdoms = ["bacteria", "viruses", "archaea", "eukaryota", null];
+
+const partition = (data, radius, attribute) => {
+  const h = hierarchy(data).sort((a, b) => b[attribute] - a[attribute]);
+  return d3partition().size([2 * Math.PI, radius])(h);
+};
+
+const getValue = (d, attributeName) => {
+  if (d[attributeName]) return d[attributeName];
+  if (!d.children) return 0;
+  return d.children.reduce(
+    (agg, item) => agg + getValue(item, attributeName),
+    0
+  );
+};
+
+const prepareTreeData = (node, depth, maxDepth, weightAttribute) => {
+  if (!node) return;
+  /* eslint-disable no-param-reassign */
+
+  node.value = node?.[weightAttribute];
+
+  if (depth >= maxDepth && node?.children?.length) {
+    node._children = node.children;
+    node.children = null;
+  } else if (node?._children?.length) {
+    node.children = node._children;
+    node._children = null;
+  }
+  /* eslint-enable no-param-reassign */
+
+  if (node?.children?.length) {
+    for (const child of node.children) {
+      prepareTreeData(child, depth + 1, maxDepth, weightAttribute);
+    }
+  }
+};
+const getLineageFromNode = (node, attributeName, attributeID) => {
+  if (!node.parent) {
+    return [
+      {
+        name: node.data[attributeName],
+        id: node.data[attributeID],
+      },
+    ];
+  }
+  return [
+    ...getLineageFromNode(node.parent, attributeName, attributeID),
+    {
+      name: node.data[attributeName],
+      id: node.data[attributeID],
+    },
+  ];
+};
+
+const getDistanceOfPointsInRadians = (point1, point2) => {
+  const angle = Math.abs(point1.angle - point2.angle) / 2;
+  return 2 * point1.radius * Math.sin(angle);
+};
+
+class NightingaleSunburst extends LitElement {
+  static properties = {
+    side: { type: Number },
+    "weight-attribute": { type: String },
+    "weight-attribute-label": { type: String },
+    "name-attribute": { type: String },
+    "id-attribute": { type: String },
+    "max-depth": { type: Number },
+    "font-size": { type: Number },
+    "show-label": { type: Boolean },
+  };
+
+  constructor() {
+    super();
+    this.side = 100;
+    this["weight-attribute"] = "value";
+    this["weight-attribute-label"] = "Value";
+    this["name-attribute"] = "name";
+    this["id-attribute"] = "id";
+    this["max-depth"] = Infinity;
+    this["font-size"] = 10;
+    this.activeSegment = null;
+    this.holdSegment = false;
+    this.topOptions = superkingdoms;
+  }
+
+  prepareTree() {
+    if (!this._data) return;
+    prepareTreeData(this._data, 0, this["max-depth"], this["weight-attribute"]);
+    this.root = partition(this._data, this.side / 2, this["weight-attribute"]);
```

Comment:
Ah, I see, so what is used to draw is root, so prepare data could return a clone of the data with the modifications instead of a mutated object, and that could be used here to pass to partition()

Reply:
Cool, I'll work on a version of prepare data, that clones the original data.